### PR TITLE
stream: extract thread safe RTP receiver

### PR DESCRIPTION
**Pull request by cspiel1**
This is the first part of https://github.com/baresip/baresip/pull/2454.

In a future PR we'll add a configuration option that enables processing of incoming RTP packets into a
separate thread, the RX thread.

This PR only does the refactoring.

- Move the fields from `struct receiver` mostly used by RX thread to rtprecv.c
    and rename to `struct rtp_receiver`. The fields only used by main thread
    are collected in `struct rxmain`.
- stream.c `rtp_handler()` --> rtprecv.c `rtprecv_decode()`
- stream.c `handle_rtp()` --> rtprecv.c
- stream.c `rtcp_handler()` --> rtprecv.c `rtprecv_handle_rtcp()`
- stream.c `mnat_connected_handler()` --> rtprecv.c
    `rtprecv_mnat_connected_handler()`
- stream.c `recv_set_ssrc() --> rtprecv.c `rtprecv_set_ssrc()`

TODOs:

- [x] test if audio/video works
- [x] check thread id of `pass_..._work` and `async_work_main`
- [x] test if rtcp works
- [x] test if inbound DTMF works
- [x] test if rtp-establish events work
- [x] test if mnat connected events work - at least **ice** works

### CODE REVIEW

File: src/rtprecv.c [hunk 1]
```diff
@@ -0,0 +1,746 @@
...
+	bool first = false;
+	int err = 0;
+
+	MAGIC_CHECK(rx);
+	if (!rx)
+		return;
+
+	mtx_lock(rx->mtx);
+	if (!rx->enabled)
+		goto unlock;
```
Comment (sreimers) [inline src/rtprecv.c:330]: It's more readable to unlock here and return, instead of a goto

File: src/rtprecv.c [hunk 2]
```diff
@@ -0,0 +1,746 @@
...
+
+	if (rtp_pt_is_rtcp(hdr->pt)) {
+		debug("stream: drop incoming RTCP packet on RTP port"
+		     " (pt=%u)\n", hdr->pt);
+		err = ENOENT;
+		goto unlock;
```
Comment (sreimers) [inline src/rtprecv.c:336]: It's more readable to unlock here and return, instead of a goto

File: src/rtprecv.c [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+	}
+	mtx_unlock(rx->mtx);
+
+	/* payload-type changed? */
+	if (hdr->pt != rx->pt) {
+		rx->pt = hdr->pt;
+
+		err = pass_pt_work(rx, hdr->pt, mb);
+		if (err && err != ENODATA)
+			goto out;
```
Comment (sreimers) [inline src/rtprecv.c:379]: You can simply return here.

File: src/rtprecv.c [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+}
+
+
+static void rtprecv_check_stop(void *arg)
+{
+	struct rtp_receiver *rx = arg;
+	bool run;
+
+	mtx_lock(rx->mtx);
+	run = rx->run;
```
Comment (sreimers) [inline src/rtprecv.c:163]: Can `rx->run` be atomic? it should be easier to handle all these cases.

Comment (cspiel1) [reply src/rtprecv.c:163]: We need the `mtx` anyway. But function `rtprecv_running()` reads `rx->run` without holding the lock. Thus it should be atomic which in this case is better than locking the mutex.

File: src/rtprecv.c [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+	bool join = false;
+
+	mtx_lock(rx->mtx);
+	if (re_atomic_rlx(&rx->run)) {
+		join = true;
+		re_atomic_rlx_set(&rx->run, false);
+	}
+	mtx_unlock(rx->mtx);
+	if (join)
+		thrd_join(rx->thr, NULL);
```
Comment (sreimers) [inline src/rtprecv.c:565]: I think this can be simplified now (do not commit the suggestion, only used for nice diff):

```suggestion
	if (re_atomic_rlx(&rx->run)) {
		re_atomic_rlx_set(&rx->run, false);
		thrd_join(rx->thr, NULL);
	}
```

Comment (cspiel1) [reply src/rtprecv.c:565]: I think if the thread terminates before `thrd_join()` then we run into troubles. We had such problems already, right?

Comment (sreimers) [reply src/rtprecv.c]: A exited thread is fine (man pthread_join):
```
The  pthread_join() function waits for the thread specified by thread to terminate.  
If that thread has already terminated, then pthread_join() returns immediately. 
```
Only two concurrent joins are problematic.

Comment (sreimers) [reply src/rtprecv.c]: It's only problmatic if `thrd_join()` is missing and not called.

Comment (cspiel1) [reply src/rtprecv.c]: ok, thanks!

**Comment by cspiel1** [comment]
Maybe we should test ice `mnat_connected_h` callback. In the code of this PR it is function `rtprecv_mnat_connected_handler()`. So if somebody have a setup for ice, please let me know!

File: src/stream.c [hunk 6]
```diff
@@ -240,47 +217,102 @@ int stream_enable_tx(struct stream *strm, bool enable)
...
 	strm->terminated = true;
-	stream_enable_tx(strm, false);
+	stream_enable(strm, false);
 	strm->errorh = NULL;
-	jbuf_flush(strm->rx.jbuf);
 
+	strm->rx = mem_deref(strm->rx);
 	if (errorh)
 		errorh(strm, err, strm->sess_arg);
+
```
Comment (cspiel1) [inline src/stream.c:289]: unrelated

**Comment by cspiel1** [comment]
Any more suggestions/comments?

**Comment by sreimers** [comment]
I did a quick baresip-webrtc test and get many (in both rx modes):

```
rtcp_sess: could not add member: 0xe89878b8      
stream: drop incoming RTCP packet on RTP port (pt=72)                                                                                                       
rtcp_sess: could not add member: 0xe89878b9
stream: drop incoming RTCP packet on RTP port (pt=72)
rtcp_sess: could not add member: 0xe89878b9
stream: drop incoming RTCP packet on RTP port (pt=72)       
rtcp_sess: could not add member: 0xe89878b9         
stream: drop incoming RTCP packet on RTP port (pt=72)                         
stream: drop incoming RTCP packet on RTP port (pt=73)                         
rtcp_sess: could not add member: 0xe89878b9                        
stream: drop incoming RTCP packet on RTP port (pt=72)                         
stream: drop incoming RTCP packet on RTP port (pt=73)                         
rtcp_sess: could not add member: 0xe89878b9                                   
stream: drop incoming RTCP packet on RTP port (pt=72)                         
rtcp_sess: could not add member: 0xe89878ba                                   
stream: drop incoming RTCP packet on RTP port (pt=72)
rtcp_sess: could not add member: 0xe89878ba                                   
stream: drop incoming RTCP packet on RTP port (pt=72)                                                                                                       
rtcp_sess: could not add member: 0xe89878ba                                   
stream: drop incoming RTCP packet on RTP port (pt=72)                         
rtcp_sess: could not add member: 0xe89878ba                                   
stream: drop incoming RTCP packet on RTP port (pt=72)        
stream: drop incoming RTCP packet on RTP port (pt=73)
stream: drop incoming RTCP packet on RTP port (pt=72)
```

**Comment by cspiel1** [comment]
`baresip-webrtc` does not compile.

I'll try to reproduce this with baresip account `rtcp_mux=yes`.

**Comment by cspiel1** [comment]
Here is the fix. I'll also add a unit test or a check in `test_call_webrtc()`.

**Comment by sreimers** [comment]
Ok looks good now, thanks.

> `baresip-webrtc` does not compile.

Did you try to compile the old archived repo? It's included within baresip now: build/webrtc/baresip-webrtc

We have a data race with `ice`:
```bash
  Write of size 1 at 0x7b540002001b by main thread (mutexes: write M0):                                                                                                                                                                                                                                                 
    #0 conncheck_handler baresip/modules/ice/ice.c:732:15 (baresip-webrtc+0x162b69) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #1 icem_checklist_update re/src/ice/chklist.c:285:3 (baresip-webrtc+0x23a57d) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #2 pace_next re/src/ice/connchk.c:34:2 (baresip-webrtc+0x23d5b2) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #3 stunc_resp_handler re/src/ice/connchk.c:202:2 (baresip-webrtc+0x23cd16) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #4 completed re/src/stun/ctrans.c:64:3 (baresip-webrtc+0x26be90) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #5 stun_ctrans_recv re/src/stun/ctrans.c:216:3 (baresip-webrtc+0x26bb59) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #6 helper_recv_handler re/src/ice/comp.c:55:10 (baresip-webrtc+0x23b364) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #7 udp_read re/src/udp/udp.c:195:10 (baresip-webrtc+0x281e0d) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)          
    #8 udp_read_handler re/src/udp/udp.c:213:2 (baresip-webrtc+0x2809a6) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #9 fd_poll re/src/main/main.c:802:4 (baresip-webrtc+0x246d02) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)        
    #10 re_main re/src/main/main.c:976:9 (baresip-webrtc+0x245f96) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)         
    #11 main baresip/webrtc/src/main.c:223:2 (baresip-webrtc+0x11f3a3) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
                                                                                                                                                                                                                                                                                                                        
  Previous read of size 1 at 0x7b540002001b by thread T13 (mutexes: write M1):                                                                              
    #0 all_completed baresip/modules/ice/ice.c:671:12 (baresip-webrtc+0x162ed5) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #1 conncheck_handler baresip/modules/ice/ice.c:745:19 (baresip-webrtc+0x162c9d) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #2 icem_checklist_update re/src/ice/chklist.c:285:3 (baresip-webrtc+0x23a57d) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #3 pace_next re/src/ice/connchk.c:34:2 (baresip-webrtc+0x23d5b2) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)            
    #4 stunc_resp_handler re/src/ice/connchk.c:202:2 (baresip-webrtc+0x23cd16) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)   
    #5 completed re/src/stun/ctrans.c:64:3 (baresip-webrtc+0x26be90) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)             
    #6 stun_ctrans_recv re/src/stun/ctrans.c:216:3 (baresip-webrtc+0x26bb59) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #7 helper_recv_handler re/src/ice/comp.c:55:10 (baresip-webrtc+0x23b364) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #8 udp_read re/src/udp/udp.c:195:10 (baresip-webrtc+0x281e0d) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)    
    #9 udp_read_handler re/src/udp/udp.c:213:2 (baresip-webrtc+0x2809a6) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #10 fd_poll re/src/main/main.c:802:4 (baresip-webrtc+0x246d02) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #11 re_main re/src/main/main.c:976:9 (baresip-webrtc+0x245f96) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #12 rtprecv_thread baresip/src/rtprecv.c:185:8 (baresip-webrtc+0x2039cb) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #13 handler re/src/thread/thread.c:101:9 (baresip-webrtc+0x27998e) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #14 handler re/src/thread/posix.c:23:27 (baresip-webrtc+0x2c280e) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    
  Location is heap block of size 1016 at 0x7b6000002000 allocated by main thread:           
    #0 malloc <null> (baresip-webrtc+0xb1013) (BuildId:
    #1 mem_alloc re/src/mem/mem.c:166:6 (baresip-webrtc+0x24ad4b) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)              
    #2 mem_zalloc re/src/mem/mem.c:198:6 (baresip-webrtc+0x24b145) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)     
    #3 stream_alloc baresip/src/stream.c:541:6 (baresip-webrtc+0x136eba) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)      
    #4 audio_alloc baresip/src/audio.c:1083:8 (baresip-webrtc+0x1e469f) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)      
    #5 peerconnection_add_audio_track baresip/src/peerconn.c:326:8 (baresip-webrtc+0x131911) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)          
    #6 session_start baresip/webrtc/src/sess.c:154:8 (baresip-webrtc+0x11f652) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)          
    #7 http_req_handler baresip/webrtc/src/demo.c:156:9 (baresip-webrtc+0x11e01e) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #8 recv_handler re/src/http/server.c:281:4 (baresip-webrtc+0x2356bb) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #9 tcp_recv_handler re/src/tcp/tcp.c:442:3 (baresip-webrtc+0x2752ef) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #10 fd_poll re/src/main/main.c:802:4 (baresip-webrtc+0x246d02) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #11 re_main re/src/main/main.c:976:9 (baresip-webrtc+0x245f96) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #12 main baresip/webrtc/src/main.c:223:2 (baresip-webrtc+0x11f3a3) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)

  Mutex M0 (0x7b3c000079b0) created at:
    #0 pthread_mutex_init <null> (baresip-webrtc+0xd1428) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #1 mtx_init re/src/thread/posix.c:168:2 (baresip-webrtc+0x2c2db5) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #2 _mutex_alloc re/src/thread/thread.c:51:8 (baresip-webrtc+0x2796b1) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #3 mutex_alloc_tp re/src/thread/thread.c:77:9 (baresip-webrtc+0x279771) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #4 re_alloc re/src/main/main.c:144:8 (baresip-webrtc+0x244dd4) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #5 re_thread_init re/src/main/main.c:1170:8 (baresip-webrtc+0x247768) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #6 rtprecv_thread baresip/src/rtprecv.c:174:2 (baresip-webrtc+0x203919) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #7 handler re/src/thread/thread.c:101:9 (baresip-webrtc+0x27998e) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #8 handler re/src/thread/posix.c:23:27 (baresip-webrtc+0x2c280e) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)

  Mutex M1 (0x7b3c000002a0) created at:
    #0 pthread_mutex_init <null> (baresip-webrtc+0xd1428) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #1 mtx_init re/src/thread/posix.c:162:3 (baresip-webrtc+0x2c2d85) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #2 _mutex_alloc re/src/thread/thread.c:51:8 (baresip-webrtc+0x2796b1) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #3 mutex_alloc re/src/thread/thread.c:71:9 (baresip-webrtc+0x279627) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #4 tmrl_alloc re/src/tmr/tmr.c:70:8 (baresip-webrtc+0x279a82) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #5 re_alloc re/src/main/main.c:151:8 (baresip-webrtc+0x244e40) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #6 re_thread_init re/src/main/main.c:1170:8 (baresip-webrtc+0x247768) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #7 libre_init re/src/main/init.c:170:8 (baresip-webrtc+0x244b10) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #8 main baresip/webrtc/src/main.c:153:8 (baresip-webrtc+0x11ef13) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)

  Thread T13 'RX thread' (tid=37000, running) created by main thread at:
    #0 pthread_create <null> (baresip-webrtc+0x9bee6) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #1 thrd_create re/src/thread/posix.c:42:8 (baresip-webrtc+0x2c277c) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #2 thread_create_name re/src/thread/thread.c:123:8 (baresip-webrtc+0x2798d3) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #3 rtprecv_start_thread baresip/src/rtprecv.c:654:8 (baresip-webrtc+0x20385f) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #4 stream_start_receiver baresip/src/stream.c:230:2 (baresip-webrtc+0x136308) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #5 tmr_poll re/src/tmr/tmr.c:160:3 (baresip-webrtc+0x279cee) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #6 re_main re/src/main/main.c:996:3 (baresip-webrtc+0x245fd9) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
    #7 main baresip/webrtc/src/main.c:223:2 (baresip-webrtc+0x11f3a3) (BuildId: 34569428fb4e7aee53fcff5350fafe3b41527bb8)
```

At first look it's a bit strange that stun answers are received by `main` and `RX thread` poll loop.

**Comment by cspiel1** [comment]
Yes, indeed I tried to compile the old repo.

I added a unit test. But with rtcp mux the first RTCP packet which is the RTCP_APP packet is not received. This has to be fixed.

Thanks for the sanitizer warnings! This PR is only the first part for refactoring stream.c. The refactoring for audio.c will follow in a second PR. For sanitizer checks I have to rebase the branch `rx_thread_stream_receiver` onto this branch (`rx_thread_1`).

I'll try this soon.

**Comment by cspiel1** [comment]
Also in `main` such a unit test fails. Here a PR: https://github.com/baresip/baresip/pull/2692

**Comment by cspiel1** [comment]
Another note to the sanitizer warnings:

Without the second part of `rx_thread_stream_receiver` branch, which makes audio.c thread safe the RX thread mode is not thread safe. Should we add a warning if RX thread mode is configured? Or disable completely or remove the new setting and just keep the refactoring?

**Comment by sreimers** [comment]
> Should we add a warning if RX thread mode is configured?

Yes a warning is good, or maybe we should hide the option for now. and we should try to keep a wiki page with tested/not tested modules and known problems etc.

**Comment by alfredh** [comment]
this PR contains two different things:

1. refactoring
2. thread feature


It would be better and less intrusive to make 1 PR that only has refactoring first.

**Comment by cspiel1** [comment]
Okay, I remove the config and the rx thread for this PR.

**Comment by cspiel1** [comment]
Updated the description. Now only refactoring is contained and mentioned.

File: src/rtprecv.c [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+
+	MAGIC_INIT(rx);
+	rx->strm   = strm;
+	rx->rtph   = rtph;
+	rx->pth    = pth;
+	rx->arg    = arg;
+	rx->pseq   = -1;
+	rx->pt     = -1;
+	err  = str_dup(&rx->name, name);
+	err |= mutex_alloc(&rx->mtx);
```
Comment (sreimers) [inline src/rtprecv.c:447]: Here is a `if (err) goto out;` missing.

File: src/rtprecv.c [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+	}
+
+	rx->metric = metric_alloc();
+	if (!rx->metric)
+		err |= ENOMEM;
+	else
+		err |= metric_init(rx->metric);
+
+	if (err)
+		goto out;
```
Comment (sreimers) [inline src/rtprecv.c:474]: Not needed can simply fall through.

File: src/rtprecv.c [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+		err |= jbuf_set_type(rx->jbuf, cfg->audio.jbtype);
+	}
+
+	/* Video Jitter buffer */
+	if (stream_type(strm) == MEDIA_VIDEO &&
+	    cfg->video.jbtype != JBUF_OFF && cfg->video.jbuf_del.max) {
+
+		err = jbuf_alloc(&rx->jbuf, cfg->video.jbuf_del.min,
+				 cfg->video.jbuf_del.max);
+		err |= jbuf_set_type(rx->jbuf, cfg->video.jbtype);
```
Comment (sreimers) [inline src/rtprecv.c:464]: You should add a explicit `if (err) goto out;` right after `jbuf_alloc`, otherwise you get `ENOMEM(12) | EINVAL(22)` = misleading EROFS(30) errors. jbuf_set_type return code can be ignored.

File: src/core.h [hunk 10]
```diff
@@ -451,3 +454,32 @@ struct media_track *mediatrack_lookup_media(const struct list *medial,
...
+			  stream_rtpestab_h *rtpestabh, void *arg);
+struct metric *rtprecv_metric(struct rtp_receiver *rx);
+struct jbuf *rtprecv_jbuf(struct rtp_receiver *rx);
+void rtprecv_decode(const struct sa *src, const struct rtp_header *hdr,
+		    struct mbuf *mb, void *arg);
+void rtprecv_handle_rtcp(const struct sa *src, struct rtcp_msg *msg,
+			 void *arg);
+void rtprecv_set_socket(struct rtp_receiver *rx, struct rtp_sock *rtp);
+void rtprecv_set_ssrc(struct rtp_receiver *rx, uint32_t ssrc);
+uint64_t rtprecv_ts_last(struct rtp_receiver *rx);
```
Comment (alfredh) [inline src/core.h:477]: can input param be `const` ?

File: src/rtprecv.c [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+
+	if (err2 == EAGAIN)
+		return err2;
+
+	return err;
+}
+
+
+void rtprecv_decode(const struct sa *src, const struct rtp_header *hdr,
+		     struct mbuf *mb, void *arg)
```
Comment (alfredh) [inline src/rtprecv.c:167]: if this is a callback handler, we should try to make them static and not expose it in api.

File: src/rtprecv.c [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+	struct rtp_receiver *rx = arg;
+
+	MAGIC_CHECK(rx);
+
+	stream_mnat_connected(rx->strm, raddr1, raddr2);
+}
+
+
+void rtprecv_set_socket(struct rtp_receiver *rx, struct rtp_sock *rtp)
+{
```
Comment (alfredh) [inline src/rtprecv.c:293]: check input params here

File: src/rtprecv.c [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+	else {
+		debug("stream: receive: setting SSRC: %x\n", ssrc);
+		rx->ssrc = ssrc;
+		rx->ssrc_set = true;
+	}
+	mtx_unlock(rx->mtx);
+}
+
+
+uint64_t rtprecv_ts_last(struct rtp_receiver *rx)
```
Comment (alfredh) [inline src/rtprecv.c:319]: check input param

**Comment by alfredh** [comment]
it looks better now ...

I did a quick test with WebRTC and it appears to be still working ...

perhaps someone else could also test on a live system ?

**Comment by cspiel1** [comment]
We put this live now on our devices.

**Comment by cspiel1** [comment]
A memory increase has to be investigated.

**Comment by alfredh** [comment]
Perhaps as @juha-h to test it ?

**Comment by cspiel1** [comment]
> Looks like a big change. PR description reads: A new configuration option enables processing of incoming RTP packets into a separate thread, the RX thread. When I looked "Files changed: 5", I didn't find any configuration option,

I updated the description. The RX thread and the new config will follow in a second PR.

Please could you test this with your Android APP?

**Comment by juha-h** [comment]
@cspiel1 So right now branch https://github.com/cspiel1/baresip/tree/rx_thread_1 only does refactoring and there is no change in behavior (new thread)?  I can give it a try once my re-INVITE ACK issue has been solved.

**Comment by cspiel1** [comment]
Yes, only refactoring. Thanks!

**Comment by juha-h** [comment]
I built libbaresip-android using  baresip from https://github.com/cspiel1/baresip/tree/rx_thread_1 and my Android app using the resulting lib.  The builds went fine.  Then I made a few test calls and didn't have any problems with their audio.

**Comment by alfredh** [comment]
Thanks for testing. Looks like we can merge if we all agree ?

**Comment by juha-h** [comment]
As I wrote elsewhere, this PR causes crash when baresip receives and handles OPTIONS request (at least one with sdp body).
